File: ios/RnExpoEspTouchModule.swift
import ExpoModulesCore
import ESPTouchV2

public class RnExpoEspTouchModule: Module {
  private var espTouchTask: ESPTouchTask?
  private var isTaskRunning = false
  
  public func definition() -> ModuleDefinition {
    Name("RnExpoEspTouch")
    
    Events("onDeviceFound", "onTaskComplete", "onError")
    
    AsyncFunction("startProvisioning") { (config: ESPTouchConfig, promise: Promise) in
      guard !isTaskRunning else {
        promise.reject("TASK_RUNNING", "ESP Touch task is already running")
        return
      }
      
      let ssid = config.ssid
      let password = config.password
      let bssid = config.bssid
      let taskCount = config.taskCount ?? 1
      
      isTaskRunning = true
      
      DispatchQueue.global(qos: .background).async { [weak self] in
        guard let self = self else { return }
        
        do {
          // Create ESP Touch task
          self.espTouchTask = ESPTouchTask(
            apSsid: ssid,
            andApBssid: bssid,
            andApPassword: password
          )
          
          // Execute task
          let results = self.espTouchTask?.executeForResults(taskCount) ?? []
          
          var devices: [[String: Any]] = []
          var successCount = 0
          
          for result in results {
            if result.isSuc() {
              successCount += 1
              let device: [String: Any] = [
                "bssid": result.getBssid() ?? "",
                "ipAddress": result.getInetAddress()?.description ?? ""
              ]
              devices.append(device)
              
              // Emit device found event
              self.sendEvent("onDeviceFound", [
                "device": device
              ])
            }
          }
          
          let response: [String: Any] = [
            "success": successCount > 0,
            "devices": devices,
            "deviceCount": successCount,
            "message": successCount > 0 ? "Successfully configured \(successCount) device(s)" : "No devices configured"
          ]
          
          DispatchQueue.main.async {
            self.isTaskRunning = false
            self.sendEvent("onTaskComplete", response)
            promise.resolve(response)
          }
          
        } catch {
          DispatchQueue.main.async {
            self.isTaskRunning = false
            let errorMessage = error.localizedDescription
            self.sendEvent("onError", ["error": errorMessage])
            promise.reject("ESP_TOUCH_ERROR", errorMessage)
          }
        }
      }
    }
    
    Function("stopProvisioning") { [weak self] in
      guard let self = self else { return }
      
      if let task = self.espTouchTask {
        task.interrupt()
        self.espTouchTask = nil
      }
      self.isTaskRunning = false
    }
    
    Function("isTaskRunning") { [weak self] () -> Bool in
      return self?.isTaskRunning ?? false
    }
    
    Function("isSupported") { () -> Bool in
      return true // ESP Touch is supported on iOS
    }
  }
}

// MARK: - Configuration Struct
struct ESPTouchConfig: Record {
  @Field var ssid: String
  @Field var password: String
  @Field var bssid: String?
  @Field var taskCount: Int?
}

File: android/src/main/java/expo/modules/rnexpoesptouch/RnExpoEspTouchModule.kt
package expo.modules.rnexpoesptouch

import android.content.Context
import android.net.wifi.WifiManager
import android.os.AsyncTask
import androidx.core.os.bundleOf
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

import com.espressif.iot.esptouch.EsptouchTask
import com.espressif.iot.esptouch.IEsptouchResult
import com.espressif.iot.esptouch.IEsptouchTask

class RnExpoEspTouchModule : Module() {
  private var esptouchTask: IEsptouchTask? = null
  private var isTaskRunning = false

  override fun definition() = ModuleDefinition {
    Name("RnExpoEspTouch")

    Events("onDeviceFound", "onTaskComplete", "onError")

    AsyncFunction("startProvisioning") { config: ESPTouchConfig, promise: Promise ->
      if (isTaskRunning) {
        promise.reject("TASK_RUNNING", "ESP Touch task is already running", null)
        return@AsyncFunction
      }

      val ssid = config.ssid
      val password = config.password
      val bssid = config.bssid
      val taskCount = config.taskCount ?: 1

      isTaskRunning = true

      // Stop any existing task
      esptouchTask?.interrupt()

      // Create new ESP Touch task
      esptouchTask = EsptouchTask(ssid, bssid, password, appContext.reactContext)

      // Execute task asynchronously
      object : AsyncTask<Void, Void, List<IEsptouchResult>>() {
        override fun doInBackground(vararg params: Void?): List<IEsptouchResult> {
          return try {
            esptouchTask?.executeForResults(taskCount) ?: emptyList()
          } catch (e: Exception) {
            sendEvent("onError", bundleOf("error" to e.message))
            emptyList()
          }
        }

        override fun onPostExecute(results: List<IEsptouchResult>) {
          val devices = mutableListOf<Map<String, String>>()
          var successCount = 0

          for (result in results) {
            if (result.isSuc) {
              successCount++
              val device = mapOf(
                "bssid" to result.bssid,
                "ipAddress" to result.inetAddress.hostAddress
              )
              devices.add(device)

              // Emit device found event
              sendEvent("onDeviceFound", bundleOf("device" to device))
            }
          }

          val response = mapOf(
            "success" to (successCount > 0),
            "devices" to devices,
            "deviceCount" to successCount,
            "message" to if (successCount > 0) {
              "Successfully configured $successCount device(s)"
            } else {
              "No devices configured"
            }
          )

          isTaskRunning = false
          sendEvent("onTaskComplete", bundleOf(
            "success" to response["success"],
            "devices" to response["devices"],
            "deviceCount" to response["deviceCount"],
            "message" to response["message"]
          ))
          promise.resolve(response)
        }

        override fun onCancelled() {
          isTaskRunning = false
          val response = mapOf(
            "success" to false,
            "devices" to emptyList<Map<String, String>>(),
            "deviceCount" to 0,
            "message" to "Task was cancelled"
          )
          promise.resolve(response)
        }
      }.execute()
    }

    Function("stopProvisioning") {
      esptouchTask?.interrupt()
      esptouchTask = null
      isTaskRunning = false
    }

    Function("isTaskRunning") {
      return@Function isTaskRunning
    }

    Function("isSupported") {
      val wifiManager = appContext.reactContext?.getSystemService(Context.WIFI_SERVICE) as? WifiManager
      return@Function wifiManager?.isWifiEnabled == true
    }
  }
}

class ESPTouchConfig : Record {
  @Field val ssid: String = ""
  @Field val password: String = ""
  @Field val bssid: String? = null
  @Field val taskCount: Int? = null
}